File: packages/agent/agent/lib/dates.ts
/**
 * Datas no fuso de quem usa.
 *
 * O cron da Vercel é avaliado em **UTC**. Um schedule que roda à meia-noite
 * UTC acorda às 21h do dia anterior em São Paulo — então "hoje", para decidir
 * se um vencimento está a três dias, precisa ser calculado aqui, no código, e
 * nunca deduzido do horário em que o schedule disparou.
 */
export const TIMEZONE = "America/Sao_Paulo";

/** Data de hoje em São Paulo, YYYY-MM-DD. */
export function todayInSaoPaulo(now: Date = new Date()): string {
  return new Intl.DateTimeFormat("en-CA", {
    timeZone: TIMEZONE,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).format(now);
}

/** Dias de `from` até `to`. Negativo quando `to` já passou. */
export function daysUntil(from: string, to: string): number {
  const ms = Date.parse(`${to}T00:00:00Z`) - Date.parse(`${from}T00:00:00Z`);
  return Math.round(ms / 86_400_000);
}

/**
 * Próxima ocorrência de um dia do mês, a partir de uma data.
 *
 * Meses curtos são tratados por ancoragem no último dia: um compromisso de dia
 * 31 cai em 28 ou 29 de fevereiro, e não transborda para março — transbordar
 * atrasaria o aviso justamente no mês em que ele é mais fácil de esquecer.
 */
export function nextOccurrence(after: string, dayOfMonth: number): string {
  const [year, month] = after.split("-").map(Number) as [number, number, number];
  const currentDay = Number(after.slice(8, 10));

  let targetYear = year;
  let targetMonth = month;
  if (currentDay >= dayOfMonth) {
    targetMonth += 1;
    if (targetMonth > 12) {
      targetMonth = 1;
      targetYear += 1;
    }
  }

  const lastDay = new Date(Date.UTC(targetYear, targetMonth, 0)).getUTCDate();
  const day = Math.min(dayOfMonth, lastDay);

  return `${targetYear}-${String(targetMonth).padStart(2, "0")}-${String(day).padStart(2, "0")}`;
}

File: services/agent/clara/instructions/dates.py
"""Datas no fuso de quem usa. Porta de `agent/lib/dates.ts`."""

from __future__ import annotations

import calendar
from datetime import date, datetime
from zoneinfo import ZoneInfo

TIMEZONE = "America/Sao_Paulo"


def today_in_sao_paulo(now: datetime | None = None) -> str:
    """Data de hoje em São Paulo, YYYY-MM-DD."""
    moment = now or datetime.now(tz=ZoneInfo("UTC"))
    return moment.astimezone(ZoneInfo(TIMEZONE)).strftime("%Y-%m-%d")


def month_range(month: str) -> tuple[str, str]:
    """Um mês do calendário como intervalo de datas inclusivo (`from`, `to`)."""
    year, month_number = (int(part) for part in month.split("-"))
    last_day = calendar.monthrange(year, month_number)[1]
    return f"{month}-01", f"{month}-{last_day:02d}"


def days_until(from_date: str, to_date: str) -> int:
    """Dias de `from_date` até `to_date`. Negativo quando `to_date` já passou."""
    a = date.fromisoformat(from_date)
    b = date.fromisoformat(to_date)
    return (b - a).days


def next_occurrence(after: str, day_of_month: int) -> str:
    """Próxima ocorrência de um dia do mês, a partir de `after`.

    Meses curtos são tratados por ancoragem no último dia: um compromisso de
    dia 31 cai em 28 ou 29 de fevereiro, sem transbordar para março.
    """
    year, month = (int(part) for part in after.split("-")[:2])
    current_day = int(after[8:10])

    target_year, target_month = year, month
    if current_day >= day_of_month:
        target_month += 1
        if target_month > 12:
            target_month = 1
            target_year += 1

    last_day = calendar.monthrange(target_year, target_month)[1]
    day = min(day_of_month, last_day)
    return f"{target_year}-{target_month:02d}-{day:02d}"
